Bluetooth: ATT, Conn, GATT, ASCS: Changed to LOG statements for buffer allocation

Thalley opened the pull request:
In the case of buffer allocation issues, or if there is an appropriate return value from API functions, then we should not use LOG_WRN and LOG_ERR as the caller can (and should) handle the cases based on the return value.

Thalley committed:
Bluetooth: ATT: Change LOG_ERR to DBG when failing to allocate buffer

For user-initiated requests like notifications, read and write
requests, there will be an error code, so there does not need to be
a generic LOG_ERR if these fail to be allocated.

However if the response or (indication) confirmation PDUs fail
to be allocated, these should still be logged as errors.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Conn: LOG_DBG for buffer alloc fail

Use LOG_DBG rather than LOG_WRN, as that is an expected
scenario when using K_NO_WAIT (e.g. when sending
GATT notifications), and it is better for higher layers
to determine if this is indeed an error.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: GATT: Modify some LOG_WRN/ERR to LOG_DBG

When an API function returns an appropriate error code,
LOG_DBG is better suited than LOG_WRN or LOG_ERR as the
caller can handle the return value.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ASCS: Use LOG_DBG when failing to alloc buffer for notification

Since there is a retry mechanism implement, this is not something
that the user needs to be warned about.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -32,6 +32,7 @@
 #include <zephyr/sys/atomic.h>
 #include <zephyr/sys/byteorder.h>
 #include <zephyr/sys/check.h>
+#include <zephyr/sys/clock.h>
```
Used to include `K_TIMEOUT_EQ`

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -748,7 +748,7 @@ static struct net_buf *bt_att_chan_create_pdu(struct bt_att_chan *chan, uint8_t
 	buf = bt_l2cap_create_pdu_timeout(&att_pool, 0, timeout);
 	if (!buf) {
-		LOG_ERR("Unable to allocate buffer for op 0x%02x", op);
+		LOG_DBG("Unable to allocate buffer for op 0x%02x", op);
```
Hmm, yes and no. 

If you encounter this issue, you can either implement a retry mechanism (which is a robust solution), or you can increase `CONFIG_BT_ATT_TX_COUNT` to some number that is hard to figure out, especially if you have enough services implement. For example it is entirely possible that you'd require `CONFIG_BT_ATT_TX_COUNT=1000` (or even infinite buffers) for some GATT services, as each write request may trigger 2 notifications (i.e. if we get a write request per connection interval, we will always be in a buffer deficit). 

So in some (many?) cases, simply increasing `CONFIG_BT_ATT_TX_COUNT` will work, but in many other cases, that's not a scalable solution. 

In either way, not being able to allocate a buffer for e.g. a GATT notification should not be logged as an error, as the caller can handle the returned error code. If you have suggestions for a conditional for when we should consider this a LOG_ERR, please me know

PavelVPV replied:
If this error generates too much noise for user app, another option is that internal users of `bt_att_chan_create_pdu` should print error when it fails to allocate buffer. Then it will be fine because we will only see the error if e.g. ATT wasn't able to allocate buffer for error response.

Thalley replied:
I agree, I can add that, and then users can use that information to recheck their code and/or increase `CONFIG_BT_ATT_TX_COUNT`

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1627,7 +1628,17 @@ struct net_buf *bt_conn_create_pdu_timeout(struct net_buf_pool *pool,
 	if (!buf) {
-		LOG_WRN("Unable to allocate buffer within timeout");
+		/* If the timeout is K_NO_WAIT then it's more likely that we cannot allocate a
+		 * buffer. Since that should only happen for non-critical requests like ATT
```
> Not necessarily.

Can you elaborate on what part is not necessarily? :) 

> I'd say (unless I'm missing something) that if timeout is not K_NO_WAIT, then it is a user call and then it doesn't require warning as user should get an error or NULL back.

Responses and confirmations are specifically not user calls and use `BT_ATT_TIMEOUT`:
```c
	switch (att_op_get_type(op)) {
	case ATT_RESPONSE:
	case ATT_CONFIRMATION:
		/* Use a timeout only when responding/confirming */
		timeout = BT_ATT_TIMEOUT;
		break;
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5125,7 +5125,7 @@ int bt_gatt_write_without_response_cb(struct bt_conn *conn, uint16_t handle,
 	write = net_buf_append_bytes(buf, length, data, K_NO_WAIT, NULL, NULL);
 	if (write != length) {
-		LOG_WRN("Unable to allocate length %u: only %zu written", length, write);
+		LOG_DBG("Unable to allocate length %u: only %zu written", length, write);
```
I was thinking the same actually. I think there are a couple of places where we could/should replace log statements with asserts, but wanted to keep this PR simpler and focus on solving the LOG_ERR/WRN issues first

Thalley committed:
Bluetooth: GATT: ASSERT instead of LOG_DBG for write length check

In bt_gatt_write_without_response_cb there is a check for
write != length. However since the call to bt_att_create_pdu
should ensure that this is possible, this should never happen.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3482,16 +3482,16 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
 
 		if (!attr->read) {
-			LOG_ERR("Read method not set");
+			LOG_DBG("Read method not set");
```
I still think these log messages in `bt_gatt_is_subscribed` need to stay as `LOG_ERR`. Debug log level is not enabled by default. These logs are extremely useful when developing application because as I said earlier there is no other way to know if I as a developer did something wrong or it is really just peer has not subscribed on attribute. They seem to signal a real error. The API could be better and return error code instead, but it is what it is.

What particular issue do you see with these logs?

PavelVPV replied:
> I don't see a particular issue with this one, but I also don't think it matches the description of when LOG_ERR should be used:

This doesn't match with PR description.

> To me they simply signal an invalid argument, and not an error at all. Had the return type of this function been `int`, it would have been a `-EINVAL`. I'm not sure I follow why this particular argument check should be treated different than the others? Or are you arguing that all checks in this function should be LOG_ERR?

`LOG_ERR` or `LOG_WRN` doesn't matter to me. You are correctly emphasizing that this could be `-EINVAL`. But the problem that this function doesn't return `-EINVAL`. And for me as a developer there is no way to figure out what is wrong. If I didn't have source code available, I would never guess that this function could return false because I missed to implement something and not because the attribute is not subscribed.

> That's where I think we differ in terms of how we develop things; I usually enable debug logging when I'm debugging something, so to me LOG_DBG is the obvious choice.

I think if we consider this issue from the perspective of personal preferences, there is no right answer. There are tons of places in this code where I'd do things differently. I’m not sure this is a strong argument for changing it.

PavelVPV replied:
If a function returns `-EINVAL`, I as a developer at least know that I did something wrong and now need to look what. How I look at this is a different question (looking at documentation (if there is one), enabling logs (or adding if there are no logs), looking at source code (if available), etc.). But if function returns `false`, I have even no clue if that is the real `false` or not.

Regarding `bt_le_ext_adv_get_info`, this error code could be different: https://github.com/zephyrproject-rtos/zephyr/blob/bd98c836eb0591d9cac30d1d3cefc01f33632235/subsys/bluetooth/host/adv.c#L1487-L1490
Rest are fine to me.

Thalley replied:
Fair point :) 

If it unblocks this PR, I can revert the change for this line. Please let me know if this is the only line you want reverted, or if the others should too
